src/components/multi-select/multi-select.jsx

```diff
@@ -0,0 +1,316 @@
+import React, {
+  createRef,
+  useEffect,
+  useImperativeHandle,
+  useRef,
+  useState,
+  forwardRef,
+} from 'react';
+import PropTypes from 'prop-types';
+import FormControl from '../form-control/form-control.jsx';
+import FormControlIcons from '../form-control-icons/form-control-icons.jsx';
+import Icon from '../icon/icon.jsx';
+import IconButton from '../icon-button/icon-button.jsx';
+import iconCaretDown from '../../svgs/caret-down.svg';
+import iconClear from '../../svgs/clear.svg';
+import Listbox from '../listbox/listbox.jsx';
+import ListOption from '../list-option/list-option.jsx';
+import Loader from '../loader/loader.jsx';
+import NoOptions from '../no-options/no-options.jsx';
+import TagList from '../tag-list/tag-list.jsx';
+import Tag from '../tag/tag.jsx';
+import useDropdownClose from '../../hooks/use-dropdown-close.js';
+import styles from './multi-select.css';
+
+const MultiSelect = forwardRef(function MultiSelect(props, ref) {
+  const [autocompleteValue, setAutocompleteValue] = useState('');
+  const autocompleteRef = useRef();
+  const backupRef = useRef();
+  const [expanded, setExpanded] = useState(false);
+  const selfRef = ref || backupRef;
+  const [validationMessage, setValidationMessage] = useState(props.validationMessage);
+  const [value, setValue] = useState(props.value || []);
+  const valueRefs = value.map(() => createRef());
+  const visibleOptions = getVisibleOptions();
+  const visibleOptionsRefs = visibleOptions.map(() => createRef());
+  const wrapperRef = useRef(null);
+
+  const ids = {
+    emptyMessage: `${props.id}-empty`,
+    label: `${props.id}-label`,
+    listbox: `${props.id}-listbox`,
+    textbox: `${props.id}-autocomplete`,
+  };
+
+  const classNames = {
+    container: styles.container,
+    formControlChildrenContainer: props.readOnly ? styles['form-control-children-container-readonly'] : styles['form-control-children-container'],
+    iconClear: props.readOnly ? styles['icon-clear-readonly'] : styles['icon-clear'],
+    iconsContainer: styles['icon-container'],
+    input: props.readOnly ? styles['input-readonly'] : styles.input,
+    noOptionsContainer: styles['no-options'],
+    popupContainer: styles['popup-container'],
+    tagList: styles['tag-list'],
+    ...props.classNames,
+  };
+
+  const dropdownContents = () => {
```

Comment:
Consider extracting this into its own private component.

Reply:
See this comment: https://github.com/mavenlink/design-system/pull/311#discussion_r635657618